File: server-auth-service/pkg/utils/state.go
package utils

import (
	"auth-service/pkg/global"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"strings"
	"time"
)

// StateData OAuth state参数结构
type StateData struct {
	Nonce       string `json:"nonce"`        // 防重放攻击的随机值
	AppID       string `json:"app_id"`       // 应用ID
	DeviceID    string `json:"device_id"`    // 设备ID
	RedirectURI string `json:"redirect_uri"` // 回调地址
	ReturnURL   string `json:"return_url"`   // 用户目标页面
	IAT         int64  `json:"iat"`          // 签发时间戳（秒）
	EXP         int64  `json:"exp"`          // 过期时间戳（秒）
}

// EncodeState 编码state参数
func EncodeState(data *StateData) (string, error) {
	// 验证必需字段
	if data.Nonce == "" {
		return "", errors.New("nonce不能为空")
	}
	if data.AppID == "" {
		return "", errors.New("app_id不能为空")
	}
	if data.RedirectURI == "" {
		return "", errors.New("redirect_uri不能为空")
	}

	// 设置时间戳（如果未设置）
	if data.IAT == 0 {
		data.IAT = time.Now().Unix()
	}
	if data.EXP == 0 {
		data.EXP = data.IAT + 300 // 默认5分钟过期
	}

	// JSON序列化
	jsonBytes, err := json.Marshal(data)
	if err != nil {
		return "", fmt.Errorf("JSON序列化失败: %w", err)
	}

	// Base64编码
	encoded := base64.StdEncoding.EncodeToString(jsonBytes)
	return encoded, nil
}

// DecodeState 解码state参数
func DecodeState(state string) (*StateData, error) {
	if state == "" {
		return nil, errors.New("state参数为空")
	}

	// Base64解码
	jsonBytes, err := base64.StdEncoding.DecodeString(state)
	if err != nil {
		return nil, fmt.Errorf("Base64解码失败: %w", err)
	}

	// JSON反序列化
	var data StateData
	if err := json.Unmarshal(jsonBytes, &data); err != nil {
		return nil, fmt.Errorf("JSON解析失败: %w", err)
	}

	return &data, nil
}

// ValidateState 验证state参数（包含nonce验证、过期验证、redirect_uri白名单验证）
func ValidateState(state string) (*StateData, error) {
	// 1. 解码state
	data, err := DecodeState(state)
	if err != nil {
		return nil, err
	}

	// 2. 验证必需字段
	if data.Nonce == "" {
		return nil, errors.New("state中缺少nonce")
	}
	if data.AppID == "" {
		return nil, errors.New("state中缺少app_id")
	}
	if data.RedirectURI == "" {
		return nil, errors.New("state中缺少redirect_uri")
	}

	// 3. 验证过期时间
	now := time.Now().Unix()
	if data.EXP < now {
		return nil, errors.New("state已过期")
	}

	// 4. 验证并消费nonce（防重放攻击）
	if err := ValidateAndConsumeNonce(data.Nonce); err != nil {
		return nil, fmt.Errorf("nonce验证失败: %w", err)
	}

	// 5. 验证redirect_uri白名单
	if err := ValidateRedirectURI(data.AppID, data.RedirectURI); err != nil {
		return nil, fmt.Errorf("redirect_uri验证失败: %w", err)
	}

	return data, nil
}

// ValidateAndConsumeNonce 验证并消费nonce（确保只能使用一次）
func ValidateAndConsumeNonce(nonce string) error {
	if nonce == "" {
		return errors.New("nonce为空")
	}

	key := fmt.Sprintf("oauth_nonce:%s", nonce)

	// 检查nonce是否存在
	exists, err := global.Redis.Exists(key).Result()
	if err != nil {
		return fmt.Errorf("Redis查询失败: %w", err)
	}

	if exists > 0 {
		// nonce已被使用
		return errors.New("nonce已被使用（可能是重放攻击）")
	}

	// 标记nonce为已使用（存储5分钟，与state过期时间一致）
	err = global.Redis.Set(key, "used", 5*time.Minute).Err()
	if err != nil {
		return fmt.Errorf("Redis存储失败: %w", err)
	}

	return nil
}

// ValidateRedirectURI 验证redirect_uri是否在应用的白名单中
func ValidateRedirectURI(appID, redirectURI string) error {
	if appID == "" {
		return errors.New("app_id为空")
	}
	if redirectURI == "" {
		return errors.New("redirect_uri为空")
	}

	// 从数据库查询应用的redirect_uris白名单
	var app struct {
		RedirectURIs string
	}
	err := global.DB.Table("sso_applications").
		Select("redirect_uris").
		Where("app_key = ? AND status = 1", appID).
		First(&app).Error

	if err != nil {
		return fmt.Errorf("应用不存在或已禁用: %w", err)
	}

	if app.RedirectURIs == "" {
		return errors.New("应用未配置redirect_uris白名单")
	}

	// 解析白名单（逗号分隔）
	allowedURIs := strings.Split(app.RedirectURIs, ",")
	for i := range allowedURIs {
		allowedURIs[i] = strings.TrimSpace(allowedURIs[i])
	}

	// 解析请求的redirect_uri，只比较base URL（不包含查询参数）
	parsedURI, err := url.Parse(redirectURI)
	if err != nil {
		return fmt.Errorf("redirect_uri格式错误: %w", err)
	}
	// 获取base URL（scheme + host + path）
	baseURI := parsedURI.Scheme + "://" + parsedURI.Host + parsedURI.Path

	// 验证redirect_uri是否在白名单中
	for _, allowed := range allowedURIs {
		// 也解析白名单中的URL，只比较base部分
		parsedAllowed, err := url.Parse(allowed)
		if err != nil {
			continue // 跳过格式错误的白名单项
		}
		baseAllowed := parsedAllowed.Scheme + "://" + parsedAllowed.Host + parsedAllowed.Path

		if baseAllowed == baseURI {
			return nil // 验证通过
		}
	}

	return fmt.Errorf("redirect_uri不在白名单中: %s", baseURI)
}

File: web-auth-service/src/utils/state.js
/**
 * OAuth State 参数工具函数
 */

/**
 * 生成UUID v4
 * @returns {string} UUID字符串
 */
export function generateNonce() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0
    const v = c === 'x' ? r : (r & 0x3 | 0x8)
    return v.toString(16)
  })
}

/**
 * 编码state参数
 * @param {Object} data - state数据对象
 * @param {string} data.nonce - 防重放攻击的随机值
 * @param {string} data.app_id - 应用ID
 * @param {string} data.device_id - 设备ID
 * @param {string} data.redirect_uri - 回调地址
 * @param {string} data.return_url - 用户目标页面（可选）
 * @param {number} data.iat - 签发时间戳（秒，可选）
 * @param {number} data.exp - 过期时间戳（秒，可选）
 * @returns {string} Base64编码的state字符串
 */
export function encodeState(data) {
  // 验证必需字段
  if (!data.nonce) {
    throw new Error('nonce不能为空')
  }
  if (!data.app_id) {
    throw new Error('app_id不能为空')
  }
  if (!data.device_id) {
    throw new Error('device_id不能为空')
  }
  if (!data.redirect_uri) {
    throw new Error('redirect_uri不能为空')
  }

  // 设置时间戳（如果未设置）
  const now = Math.floor(Date.now() / 1000)
  if (!data.iat) {
    data.iat = now
  }
  if (!data.exp) {
    data.exp = now + 300 // 默认5分钟过期
  }

  // JSON序列化并Base64编码
  const jsonString = JSON.stringify(data)
  const base64String = btoa(jsonString)
  
  return base64String
}

/**
 * 解码state参数
 * @param {string} state - Base64编码的state字符串
 * @returns {Object} 解码后的state数据对象
 */
export function decodeState(state) {
  if (!state) {
    throw new Error('state参数为空')
  }

  try {
    // Base64解码
    const jsonString = atob(state)
    
    // JSON解析
    const data = JSON.parse(jsonString)
    
    return data
  } catch (error) {
    throw new Error('state参数解析失败: ' + error.message)
  }
}

/**
 * 验证state参数（仅前端基础验证，完整验证在后端）
 * @param {string} state - Base64编码的state字符串
 * @returns {Object} 解码后的state数据对象
 */
export function validateState(state) {
  const data = decodeState(state)

  // 验证必需字段
  if (!data.nonce) {
    throw new Error('state中缺少nonce')
  }
  if (!data.app_id) {
    throw new Error('state中缺少app_id')
  }
  if (!data.redirect_uri) {
    throw new Error('state中缺少redirect_uri')
  }

  // 验证过期时间
  const now = Math.floor(Date.now() / 1000)
  if (data.exp && data.exp < now) {
    throw new Error('state已过期')
  }

  return data
}
